File: backtracking/78_subsets.py
# Given an integer array nums of unique elements, return all possible subsets (the power set).

# The solution set must not contain duplicate subsets. Return the solution in any order

from typing import List
nums = [1, 2, 3]
# Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]

# An inefficient brute-force approach would be to generate every combination and try to add the sorted combination as a key to a hashmap.
# That way you weed out any duplicates and are only left with unique permutations without repetition.

# Maybe there is an approach that could remove an increasing amount of elements per loop-cycle?
# First it removes none, then it removes each number once, then it removes each combination of two numbers etc. and saves the results.


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []


obj = Solution()
obj.subsets(nums)

File: backtracking/78_subsets.cs
public class Solution
{
    public IList<IList<int>> Subsets(int[] nums)
    {
        IList<int> cur = [];
        IList<IList<int>> result = FindCombinations(-1, cur, nums);
        result.Add([]);
        return result;
    }

    public IList<IList<int>> FindCombinations(int l, IList<int> cur, int[] array)
    {
        IList<IList<int>> result = [];
        l++;
        while (l < array.Length)
        {
            IList<int> copy = new List<int>(cur)
            {
                array[l]
            };
            result.Add(copy);
            IList<IList<int>> temp = FindCombinations(l, copy, array);
            foreach (IList<int> arr in temp)
            {
                result.Add(arr);
            }
            l++;
        }
        return result;
    }
}